migrations/lib/importer/steps/categories.rb

```diff
@@ -0,0 +1,251 @@
+                   auto_close_based_on_last_post
+                   auto_close_hours
+                   color
+                   created_at
+                   updated_at
+                   default_list_filter
+                   default_slow_mode_seconds
+                   default_top_period
+                   default_view
+                   description
+                   email_in
+                   email_in_allow_strangers
+                   emoji
+                   icon
+                   mailinglist_mirror
+                   minimum_required_tags
+                   name
+                   name_lower
+                   navigate_to_first_post_after_read
+                   num_featured_topics
+                   parent_category_id
+                   position
+                   read_only_banner
+                   read_restricted
+                   show_subcategory_list
+                   slug
+                   sort_ascending
+                   sort_order
+                   style_type
+                   subcategory_list_style
+                   text_color
+                   topic_featured_link_allowed
+                   topic_template
+                   uploaded_background_dark_id
+                   uploaded_background_id
+                   uploaded_logo_dark_id
+                   uploaded_logo_id
+                   user_id
+                 ]
+
+    total_rows_query <<~SQL, MappingType::CATEGORIES
+      SELECT COUNT(*)
+      FROM categories c
+          LEFT JOIN mapped.ids mapped_category
+            ON c.original_id = mapped_category.original_id AND mapped_category.type = ?
+      WHERE mapped_category.original_id IS NULL
+    SQL
+
+    rows_query <<~SQL, MappingType::CATEGORIES, MappingType::USERS, MappingType::UPLOADS
+      WITH
+          RECURSIVE
+          tree AS (
+                    SELECT c.*, 0 AS level
+                      FROM categories c
+                    WHERE c.parent_category_id IS NULL
+                    UNION ALL
+                    SELECT c.*, tree.level + 1 AS level
+                      FROM categories c,
+                          tree
+                    WHERE c.parent_category_id = tree.original_id
+                  )
+      SELECT t.*,
+            COALESCE(t.position,
+                      ROW_NUMBER() OVER (
+                        PARTITION BY parent_category_id
+                        ORDER BY parent_category_id NULLS FIRST, name
+                        )
+                    ) AS position,
+            mapped_user.discourse_id      AS mapped_user_id,
+            mapped_bg.discourse_id        AS mapped_uploaded_background_id,
+            mapped_bg_dark.discourse_id   AS mapped_uploaded_background_dark_id,
+            mapped_logo.discourse_id      AS mapped_uploaded_logo_id,
+            mapped_logo_dark.discourse_id AS mapped_uploaded_logo_dark_id
+      FROM tree t
+           LEFT JOIN mapped.ids mapped_category
```

Comment:
I like the usage of longer aliases here. Makes it much more readable. I'll copy that into other steps. :+1:

Reply:
I'd go with `mapped_categories` to fit in with table names being plural, but that's the only change I'd make. I like the pattern of having clear, readable aliases. Heck, I even experimented with not giving aliases to the IntermediateDB tables. So, `users` instead of `u`. I like it so far.